### DISCUSSION TITLE
Question about bindings

### DISCUSSION THREAD

**Discussion by Ruttie2006** [AvaloniaUI/Avalonia#9809] [Q&A] [answered]
How does the BindingExtension get the DataContext of a view?
Like, I've tried looking at the source, but I just couldn't figure it out.
Or can we not replicate such a thing?
Also, if it *is* possible, is it possible to get the view the markup extension is being added to?

**Comment by maxkatz6** [comment, marked as answer]
> Or can we not replicate such a thing?

Every markup extension used in avalonia can be replicated. Except single CompiledBinding extension which needs more compile-time information to work.

> How does the BindingExtension get the DataContext of a view?

Use `serviceProvider.GetDefaultAnchor()` method. This method will lookup XAML elements tree (how they are defined in XAML code), and search for (in a order or priority):
- Control
- IDataContextProvider (example: Application class which isn't a control, but has a DataContext)
- Or nearest IStyle object

You can see example how it is used in ReflectionBinding
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/ReflectionBindingExtension.cs#L39


> Also, if it is possible, is it possible to get the view the markup extension is being added to?

`serviceProvider.GetFirstParent<MyControl>()` or to get a root object `serviceProvider.GetService<IRootObjectProvider>()?.RootObject`
